File: db4oj.tests/src/com/db4o/db4ounit/common/soda/classes/simple/STStringTestCase.java
/* Copyright (C) 2004   db4objects Inc.   http://www.db4o.com */

package com.db4o.db4ounit.common.soda.classes.simple;
import com.db4o.*;
import com.db4o.db4ounit.common.soda.*;
import com.db4o.query.*;




public class STStringTestCase extends com.db4o.db4ounit.common.soda.util.SodaBaseTestCase implements STInterface {
	
	public String str;

	public STStringTestCase() {
	}

	public STStringTestCase(String str) {
		this.str = str;
	}

	/** needed for STInterface test */
	public Object returnSomething() {
		return str;
	}

	public Object[] createData() {
		return new Object[] {
			new STStringTestCase(null),
			new STStringTestCase("aaa"),
			new STStringTestCase("bbb"),
			new STStringTestCase("dod")};
	}
	
	public void testEquals() {
		Query q = newQuery();
		q.constrain(_array[2]);
		com.db4o.db4ounit.common.soda.util.SodaTestUtil.expectOne(q, _array[2]);
	}

	public void testNotEquals() {
		Query q = newQuery();
		Constraint c = q.constrain(_array[2]);
		q.descend("str").constraints().not();
		
		expect(q, new int[] { 0, 1, 3 });
	}

	public void testDescendantEquals() {
		Query q = newQuery();
		q.constrain(new STStringTestCase());
		q.descend("str").constrain("bbb");
		com.db4o.db4ounit.common.soda.util.SodaTestUtil.expectOne(q, new STStringTestCase("bbb"));
	}

	public void testContains() {
		Query q = newQuery();
		Constraint c = q.constrain(new STStringTestCase("od"));
		q.descend("str").constraints().contains();
		com.db4o.db4ounit.common.soda.util.SodaTestUtil.expectOne(q, new STStringTestCase("dod"));
	}

	public void testNotContains() {
		Query q = newQuery();
		Constraint c = q.constrain(new STStringTestCase("od"));
		q.descend("str").constraints().contains().not();
		com.db4o.db4ounit.common.soda.util.SodaTestUtil.expect(
			q,
			new Object[] { new STStringTestCase(null), new STStringTestCase("aaa"), new STStringTestCase("bbb")});
	}

	public void testLike() {
		Query q = newQuery();
		Constraint c = q.constrain(new STStringTestCase("do"));
		q.descend("str").constraints().like();
		com.db4o.db4ounit.common.soda.util.SodaTestUtil.expectOne(q, new STStringTestCase("dod"));
		q = newQuery();
		c = q.constrain(new STStringTestCase("od"));
		q.descend("str").constraints().like();
        
		com.db4o.db4ounit.common.soda.util.SodaTestUtil.expectOne(q,_array[3]);
	}

	public void testStartsWith() {
		Query q = newQuery();
		Constraint c = q.constrain(new STStringTestCase("do"));
		q.descend("str").constraints().startsWith(true);
		com.db4o.db4ounit.common.soda.util.SodaTestUtil.expectOne(q, new STStringTestCase("dod"));
		q = newQuery();
		c = q.constrain(new STStringTestCase("od"));
		q.descend("str").constraints().startsWith(true);
		expect(q, new int[] {});
	}

	public void testEndsWith() {
		Query q = newQuery();
		Constraint c = q.constrain(new STStringTestCase("do"));
		q.descend("str").constraints().endsWith(true);
		expect(q, new int[] {});
		q = newQuery();
		c = q.constrain(new STStringTestCase("od"));
		q.descend("str").constraints().endsWith(true);
		com.db4o.db4ounit.common.soda.util.SodaTestUtil.expectOne(q, new STStringTestCase("dod"));
		q = newQuery();
		c = q.constrain(new STStringTestCase("D"));
		q.descend("str").constraints().endsWith(false);
		com.db4o.db4ounit.common.soda.util.SodaTestUtil.expectOne(q, new STStringTestCase("dod"));
	}

	public void testNotLike() {
		Query q = newQuery();
		Constraint c = q.constrain(new STStringTestCase("aaa"));
		q.descend("str").constraints().like().not();
		com.db4o.db4ounit.common.soda.util.SodaTestUtil.expect(
			q,
			new Object[] { new STStringTestCase(null), new STStringTestCase("bbb"), new STStringTestCase("dod")});
		q = newQuery();
		c = q.constrain(new STStringTestCase("xxx"));
		q.descend("str").constraints().like();
		expect(q, new int[] {});
	}

	public void testIdentity() {
		Query q = newQuery();
		Constraint c = q.constrain(new STStringTestCase("aaa"));
		ObjectSet set = q.execute();
		STStringTestCase identityConstraint = (STStringTestCase) set.next();
		identityConstraint.str = "hihs";
		q = newQuery();
		q.constrain(identityConstraint).identity();
		identityConstraint.str = "aaa";
		com.db4o.db4ounit.common.soda.util.SodaTestUtil.expectOne(q, new STStringTestCase("aaa"));
	}

	public void testNotIdentity() {
		Query q = newQuery();
		Constraint c = q.constrain(new STStringTestCase("aaa"));
		ObjectSet set = q.execute();
		STStringTestCase identityConstraint = (STStringTestCase) set.next();
		identityConstraint.str = null;
		q = newQuery();
		q.constrain(identityConstraint).identity().not();
		identityConstraint.str = "aaa";
		com.db4o.db4ounit.common.soda.util.SodaTestUtil.expect(
			q,
			new Object[] { new STStringTestCase(null), new STStringTestCase("bbb"), new STStringTestCase("dod")});
	}

	public void testNull() {
		Query q = newQuery();
		Constraint c = q.constrain(new STStringTestCase(null));
		q.descend("str").constrain(null);
		com.db4o.db4ounit.common.soda.util.SodaTestUtil.expectOne(q, new STStringTestCase(null));
	}

	public void testNotNull() {
		Query q = newQuery();
		Constraint c = q.constrain(new STStringTestCase(null));
		q.descend("str").constrain(null).not();
		com.db4o.db4ounit.common.soda.util.SodaTestUtil.expect(
			q,
			new Object[] { new STStringTestCase("aaa"), new STStringTestCase("bbb"), new STStringTestCase("dod")});
	}

	public void testConstraints() {
		Query q = newQuery();
		q.constrain(new STStringTestCase("aaa"));
		q.constrain(new STStringTestCase("bbb"));
		Constraints cs = q.constraints();
		Constraint[] csa = cs.toArray();
		if (csa.length != 2) {
			db4ounit.Assert.fail("Constraints not returned");
		}
	}

	public void testEvaluation() {
		Query q = newQuery();
		q.constrain(new STStringTestCase(null));
		q.constrain(new Evaluation() {
			public void evaluate(Candidate candidate) {
				STStringTestCase sts = (STStringTestCase) candidate.getObject();
				candidate.include(sts.str.indexOf("od") == 1);
			}
		});
		com.db4o.db4ounit.common.soda.util.SodaTestUtil.expectOne(q, new STStringTestCase("dod"));
	}
	
    public void testCaseInsenstiveContains() {
        Query q = newQuery();
        q.constrain(STStringTestCase.class);
        q.constrain(new Evaluation() {
            public void evaluate(Candidate candidate) {
                STStringTestCase sts = (STStringTestCase) candidate.getObject();
                candidate.include(sts.str.toLowerCase().indexOf("od") >= 0);
            }
        });
        com.db4o.db4ounit.common.soda.util.SodaTestUtil.expectOne(q, new STStringTestCase("dod"));
    }

}

File: db4on/test/src/com/db4o/db4ounit/common/soda/classes/simple/STStringTestCase.cs
namespace com.db4o.db4ounit.common.soda.classes.simple
{
	public class STStringTestCase : com.db4o.db4ounit.common.soda.util.SodaBaseTestCase
		, com.db4o.db4ounit.common.soda.STInterface
	{
		public string str;

		public STStringTestCase()
		{
		}

		public STStringTestCase(string str)
		{
			this.str = str;
		}

		/// <summary>needed for STInterface test</summary>
		public virtual object ReturnSomething()
		{
			return str;
		}

		public override object[] CreateData()
		{
			return new object[] { new com.db4o.db4ounit.common.soda.classes.simple.STStringTestCase
				(null), new com.db4o.db4ounit.common.soda.classes.simple.STStringTestCase("aaa")
				, new com.db4o.db4ounit.common.soda.classes.simple.STStringTestCase("bbb"), new 
				com.db4o.db4ounit.common.soda.classes.simple.STStringTestCase("dod") };
		}

		public virtual void TestEquals()
		{
			com.db4o.query.Query q = NewQuery();
			q.Constrain(_array[2]);
			com.db4o.db4ounit.common.soda.util.SodaTestUtil.ExpectOne(q, _array[2]);
		}

		public virtual void TestNotEquals()
		{
			com.db4o.query.Query q = NewQuery();
			q.Constrain(_array[2]);
			q.Descend("str").Constraints().Not();
			Expect(q, new int[] { 0, 1, 3 });
		}

		public virtual void TestDescendantEquals()
		{
			com.db4o.query.Query q = NewQuery();
			q.Constrain(new com.db4o.db4ounit.common.soda.classes.simple.STStringTestCase());
			q.Descend("str").Constrain("bbb");
			com.db4o.db4ounit.common.soda.util.SodaTestUtil.ExpectOne(q, new com.db4o.db4ounit.common.soda.classes.simple.STStringTestCase
				("bbb"));
		}

		public virtual void TestContains()
		{
			com.db4o.query.Query q = NewQuery();
			q.Constrain(new com.db4o.db4ounit.common.soda.classes.simple.STStringTestCase("od"
				));
			q.Descend("str").Constraints().Contains();
			com.db4o.db4ounit.common.soda.util.SodaTestUtil.ExpectOne(q, new com.db4o.db4ounit.common.soda.classes.simple.STStringTestCase
				("dod"));
		}

		public virtual void TestNotContains()
		{
			com.db4o.query.Query q = NewQuery();
			q.Constrain(new com.db4o.db4ounit.common.soda.classes.simple.STStringTestCase("od"
				));
			q.Descend("str").Constraints().Contains().Not();
			com.db4o.db4ounit.common.soda.util.SodaTestUtil.Expect(q, new object[] { new com.db4o.db4ounit.common.soda.classes.simple.STStringTestCase
				(null), new com.db4o.db4ounit.common.soda.classes.simple.STStringTestCase("aaa")
				, new com.db4o.db4ounit.common.soda.classes.simple.STStringTestCase("bbb") });
		}

		public virtual void TestLike()
		{
			com.db4o.query.Query q = NewQuery();
			q.Constrain(new com.db4o.db4ounit.common.soda.classes.simple.STStringTestCase("do"
				));
			q.Descend("str").Constraints().Like();
			com.db4o.db4ounit.common.soda.util.SodaTestUtil.ExpectOne(q, new com.db4o.db4ounit.common.soda.classes.simple.STStringTestCase
				("dod"));
			q = NewQuery();
			q.Constrain(new com.db4o.db4ounit.common.soda.classes.simple.STStringTestCase("od"
				));
			q.Descend("str").Constraints().Like();
			com.db4o.db4ounit.common.soda.util.SodaTestUtil.ExpectOne(q, _array[3]);
		}

		public virtual void TestStartsWith()
		{
			com.db4o.query.Query q = NewQuery();
			q.Constrain(new com.db4o.db4ounit.common.soda.classes.simple.STStringTestCase("do"
				));
			q.Descend("str").Constraints().StartsWith(true);
			com.db4o.db4ounit.common.soda.util.SodaTestUtil.ExpectOne(q, new com.db4o.db4ounit.common.soda.classes.simple.STStringTestCase
				("dod"));
			q = NewQuery();
			q.Constrain(new com.db4o.db4ounit.common.soda.classes.simple.STStringTestCase("od"
				));
			q.Descend("str").Constraints().StartsWith(true);
			Expect(q, new int[] {  });
		}

		public virtual void TestEndsWith()
		{
			com.db4o.query.Query q = NewQuery();
			q.Constrain(new com.db4o.db4ounit.common.soda.classes.simple.STStringTestCase("do"
				));
			q.Descend("str").Constraints().EndsWith(true);
			Expect(q, new int[] {  });
			q = NewQuery();
			q.Constrain(new com.db4o.db4ounit.common.soda.classes.simple.STStringTestCase("od"
				));
			q.Descend("str").Constraints().EndsWith(true);
			com.db4o.db4ounit.common.soda.util.SodaTestUtil.ExpectOne(q, new com.db4o.db4ounit.common.soda.classes.simple.STStringTestCase
				("dod"));
			q = NewQuery();
			q.Constrain(new com.db4o.db4ounit.common.soda.classes.simple.STStringTestCase("D"
				));
			q.Descend("str").Constraints().EndsWith(false);
			com.db4o.db4ounit.common.soda.util.SodaTestUtil.ExpectOne(q, new com.db4o.db4ounit.common.soda.classes.simple.STStringTestCase
				("dod"));
		}

		public virtual void TestNotLike()
		{
			com.db4o.query.Query q = NewQuery();
			q.Constrain(new com.db4o.db4ounit.common.soda.classes.simple.STStringTestCase("aaa"
				));
			q.Descend("str").Constraints().Like().Not();
			com.db4o.db4ounit.common.soda.util.SodaTestUtil.Expect(q, new object[] { new com.db4o.db4ounit.common.soda.classes.simple.STStringTestCase
				(null), new com.db4o.db4ounit.common.soda.classes.simple.STStringTestCase("bbb")
				, new com.db4o.db4ounit.common.soda.classes.simple.STStringTestCase("dod") });
			q = NewQuery();
			q.Constrain(new com.db4o.db4ounit.common.soda.classes.simple.STStringTestCase("xxx"
				));
			q.Descend("str").Constraints().Like();
			Expect(q, new int[] {  });
		}

		public virtual void TestIdentity()
		{
			com.db4o.query.Query q = NewQuery();
			q.Constrain(new com.db4o.db4ounit.common.soda.classes.simple.STStringTestCase("aaa"
				));
			com.db4o.ObjectSet set = q.Execute();
			com.db4o.db4ounit.common.soda.classes.simple.STStringTestCase identityConstraint = 
				(com.db4o.db4ounit.common.soda.classes.simple.STStringTestCase)set.Next();
			identityConstraint.str = "hihs";
			q = NewQuery();
			q.Constrain(identityConstraint).Identity();
			identityConstraint.str = "aaa";
			com.db4o.db4ounit.common.soda.util.SodaTestUtil.ExpectOne(q, new com.db4o.db4ounit.common.soda.classes.simple.STStringTestCase
				("aaa"));
		}

		public virtual void TestNotIdentity()
		{
			com.db4o.query.Query q = NewQuery();
			q.Constrain(new com.db4o.db4ounit.common.soda.classes.simple.STStringTestCase("aaa"
				));
			com.db4o.ObjectSet set = q.Execute();
			com.db4o.db4ounit.common.soda.classes.simple.STStringTestCase identityConstraint = 
				(com.db4o.db4ounit.common.soda.classes.simple.STStringTestCase)set.Next();
			identityConstraint.str = null;
			q = NewQuery();
			q.Constrain(identityConstraint).Identity().Not();
			identityConstraint.str = "aaa";
			com.db4o.db4ounit.common.soda.util.SodaTestUtil.Expect(q, new object[] { new com.db4o.db4ounit.common.soda.classes.simple.STStringTestCase
				(null), new com.db4o.db4ounit.common.soda.classes.simple.STStringTestCase("bbb")
				, new com.db4o.db4ounit.common.soda.classes.simple.STStringTestCase("dod") });
		}

		public virtual void TestNull()
		{
			com.db4o.query.Query q = NewQuery();
			q.Constrain(new com.db4o.db4ounit.common.soda.classes.simple.STStringTestCase(null
				));
			q.Descend("str").Constrain(null);
			com.db4o.db4ounit.common.soda.util.SodaTestUtil.ExpectOne(q, new com.db4o.db4ounit.common.soda.classes.simple.STStringTestCase
				(null));
		}

		public virtual void TestNotNull()
		{
			com.db4o.query.Query q = NewQuery();
			q.Constrain(new com.db4o.db4ounit.common.soda.classes.simple.STStringTestCase(null
				));
			q.Descend("str").Constrain(null).Not();
			com.db4o.db4ounit.common.soda.util.SodaTestUtil.Expect(q, new object[] { new com.db4o.db4ounit.common.soda.classes.simple.STStringTestCase
				("aaa"), new com.db4o.db4ounit.common.soda.classes.simple.STStringTestCase("bbb"
				), new com.db4o.db4ounit.common.soda.classes.simple.STStringTestCase("dod") });
		}

		public virtual void TestConstraints()
		{
			com.db4o.query.Query q = NewQuery();
			q.Constrain(new com.db4o.db4ounit.common.soda.classes.simple.STStringTestCase("aaa"
				));
			q.Constrain(new com.db4o.db4ounit.common.soda.classes.simple.STStringTestCase("bbb"
				));
			com.db4o.query.Constraints cs = q.Constraints();
			com.db4o.query.Constraint[] csa = cs.ToArray();
			if (csa.Length != 2)
			{
				Db4oUnit.Assert.Fail("Constraints not returned");
			}
		}

		public virtual void TestEvaluation()
		{
			com.db4o.query.Query q = NewQuery();
			q.Constrain(new com.db4o.db4ounit.common.soda.classes.simple.STStringTestCase(null
				));
			q.Constrain(new _AnonymousInnerClass179(this));
			com.db4o.db4ounit.common.soda.util.SodaTestUtil.ExpectOne(q, new com.db4o.db4ounit.common.soda.classes.simple.STStringTestCase
				("dod"));
		}

		private sealed class _AnonymousInnerClass179 : com.db4o.query.Evaluation
		{
			public _AnonymousInnerClass179(STStringTestCase _enclosing)
			{
				this._enclosing = _enclosing;
			}

			public void Evaluate(com.db4o.query.Candidate candidate)
			{
				com.db4o.db4ounit.common.soda.classes.simple.STStringTestCase sts = (com.db4o.db4ounit.common.soda.classes.simple.STStringTestCase
					)candidate.GetObject();
				candidate.Include(sts.str.IndexOf("od") == 1);
			}

			private readonly STStringTestCase _enclosing;
		}

		public virtual void TestCaseInsenstiveContains()
		{
			com.db4o.query.Query q = NewQuery();
			q.Constrain(typeof(com.db4o.db4ounit.common.soda.classes.simple.STStringTestCase)
				);
			q.Constrain(new _AnonymousInnerClass191(this));
			com.db4o.db4ounit.common.soda.util.SodaTestUtil.ExpectOne(q, new com.db4o.db4ounit.common.soda.classes.simple.STStringTestCase
				("dod"));
		}

		private sealed class _AnonymousInnerClass191 : com.db4o.query.Evaluation
		{
			public _AnonymousInnerClass191(STStringTestCase _enclosing)
			{
				this._enclosing = _enclosing;
			}

			public void Evaluate(com.db4o.query.Candidate candidate)
			{
				com.db4o.db4ounit.common.soda.classes.simple.STStringTestCase sts = (com.db4o.db4ounit.common.soda.classes.simple.STStringTestCase
					)candidate.GetObject();
				candidate.Include(sts.str.ToLower().IndexOf("od") >= 0);
			}

			private readonly STStringTestCase _enclosing;
		}
	}
}
